src/Compilers/Core/Portable/Operations/IOperation.OperationList.cs

```diff
@@ -0,0 +1,140 @@
+
+            public bool Any() => Count > 0;
+
+            public IOperation First()
+            {
+                var enumerator = GetEnumerator();
+                if (enumerator.MoveNext())
+                {
+                    return enumerator.Current;
+                }
+
+                throw new InvalidOperationException();
+            }
+
+            public Reversed Reverse() => new Reversed(_operation);
+
+            public IOperation Last()
+            {
+                var enumerator = Reverse().GetEnumerator();
+                if (enumerator.MoveNext())
+                {
+                    return enumerator.Current;
+                }
+
+                throw new InvalidOperationException();
+            }
+
+            /// <summary>
+            /// Implements a struct-based enumerator for <see cref="Operation"/> nodes. This type is not hardened
+            /// to <code>default(Enumerator)</code>, and will null reference in these cases. Calling <see cref="Current"/> after
+            /// <see cref="Enumerator.MoveNext"/> has returned false will throw an <see cref="InvalidOperationException"/>.
+            /// </summary>
+            [NonDefaultable]
+            public struct Enumerator : IEnumerator<IOperation>
+            {
+                /// <summary>
+                /// Implementation of the <see cref="Enumerator.MoveNext"/> and <see cref="Enumerator.Current"/>
+                /// members are delegated to the virtual <see cref="Operation.MoveNext(int, int)"/> and
+                /// <see cref="Operation.GetCurrent(int, int)"/> methods, respectively.
+                /// </summary>
+                private readonly Operation _operation;
+                /// <summary>
+                /// 
+                /// </summary>
+                private int _currentSlot;
+                private int _currentIndex;
+
+                internal Enumerator(Operation operation)
+                {
+                    _operation = operation;
+                    _currentSlot = -1;
+                    _currentIndex = -1;
+                }
+
+                public IOperation Current
+                {
+                    get
+                    {
+                        Debug.Assert(_operation != null && _currentSlot >= 0 && _currentIndex >= 0);
+                        return _operation.GetCurrent(_currentSlot, _currentIndex);
+                    }
+                }
+
+                public bool MoveNext()
+                {
+                    bool result;
+                    (result, _currentSlot, _currentIndex) = _operation.MoveNext(_currentSlot, _currentIndex);
+                    return result;
+                }
+
+                public void Reset()
+                {
+                    _currentSlot = -1;
+                    _currentIndex = -1;
+                }
+
+                object? IEnumerator.Current => this.Current;
+                void IDisposable.Dispose() { }
```

Comment:
>void IDisposable.Dispose() { } [](http://example.com/codeflow?start=16&length=30)

Does this mean that uses of `Enumerator` will typically require a `try/finally` block and require boxing the instance as `IDisposable`?

Reply:
Actually, the compiler emits a `constrained. callvirt` to `IDisposable.Dispose()` which should avoid boxing. Sorry for the confusion.